### PR TITLE
Use std types for SFINAE usage in cms::Exception

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Replaced our own types with std types.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Utilities/interface/Exception.h [hunk 1]
```diff
@@ -226,41 +208,22 @@ namespace cms {
 
   // template <typename E>
   // inline
-  // typename detail::Desired<E, detail::is_derived_or_same<Exception,E>::value>::type &
+  // typename detail::exception_type<E> &
```
Comment (makortel) [inline FWCore/Utilities/interface/Exception.h:211]: Unrelated to this PR, but do we still need this commented out code? The comment above refers to problems with GCC 3.2.3, but we are well beyond that by now.

Comment (Dr15Jones) [reply FWCore/Utilities/interface/Exception.h:211]: If we haven't needed it for this long then it is hard to justify that we'd need it in the future.

Comment (makortel) [reply FWCore/Utilities/interface/Exception.h:211]: Could you remove them then?

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
The usual
* https://github.com/cms-sw/cmssw/issues/39803
* https://github.com/cms-sw/cmssw/issues/43293

**Comment by rappoccio** [comment]
+1

**Pull request merged**